File: Algorithms/p113_Path_Sum_II/p113_Path_Sum_II.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        if not root:
            return []
        res = []
        cur = []
        self.dfs(res,cur,root,sum)
        return res

    def dfs(self, res, cur, root, sum):
        cur.append(root.val)
        sum -= root.val
        if root.left:
            self.dfs(res,cur,root.left,sum)
        if root.right:
            self.dfs(res,cur,root.right,sum)
        if not root.left and not root.right and sum==0:
            res.append(cur[:])
        cur.pop()

File: Algorithms/p113_Path_Sum_II/p113_Path_Sum_II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        if(!root) return {};
        vector<vector<int>> res;
        vector<int> cur;
        dfs(res,cur,root,sum);
        return res;
    }
    void dfs(vector<vector<int>>& res, vector<int>& cur, TreeNode* root, int sum) {
        cur.push_back(root->val);
        sum -= root->val;
        if(root->left) dfs(res,cur,root->left,sum);
        if(root->right) dfs(res,cur,root->right,sum);
        if(!root->left&&!root->right&&sum==0) res.push_back(cur);
        cur.pop_back();
    }
};
